core/unit_test/compile_only/CompileTestDetection.cpp

```diff
@@ -0,0 +1,271 @@
+struct C {};
+
+struct MyTestCase1 {
+  using my_member_type = C;
+
+  static constexpr auto my_static_data_member = int(42);
+
+  KOKKOS_FUNCTION
+  void my_method();
+
+  KOKKOS_FUNCTION
+  int my_int_method(double, float);
+
+  KOKKOS_FUNCTION
+  double my_method_with_overloads();
+
+  KOKKOS_FUNCTION
+  B& my_method_with_overloads(my_member_type);
+
+  KOKKOS_FUNCTION
+  void my_method_with_overloads(int**);
+
+  KOKKOS_FUNCTION
+  my_member_type my_method_with_overloads(my_member_type, double);
+
+  KOKKOS_FUNCTION
+  MyTestCase1& my_method_with_overloads(A const&);
+};
+
+// TODO test these in other scopes that could trip up compilers, like function
+// scope or dependent class template scope
+
+// hopefully this isn't too lazy; it should make the tests a bit more readable
+using namespace Kokkos::Impl;
+
+//==============================================================================
+// <editor-fold desc="Test member type detection"> {{{1
+
+// template <class T>
+// using _my_member_type_archetype = typename T::my_member_type;
+KOKKOS_DECLARE_DETECTION_ARCHETYPE_1PARAM(_my_member_type_archetype, T,
+                                          typename T::my_member_type);
+
+KOKKOS_STATIC_TEST(is_detected<_my_member_type_archetype, MyTestCase1>::value);
+
+KOKKOS_STATIC_TEST(
+    is_detected_exact<C, _my_member_type_archetype, MyTestCase1>::value);
+
+KOKKOS_STATIC_TEST(
+    is_detected_convertible<C, _my_member_type_archetype, MyTestCase1>::value);
+
+// </editor-fold> end Test member type detection }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="Test method detection"> {{{1
+
+// template <class T>
+// using _my_method_archetype_1 = decltype(T{}.my_method());
+KOKKOS_DECLARE_DETECTION_ARCHETYPE_1PARAM(_my_method_archetype_1, T,
+                                          decltype(T{}.my_method()));
+
+// template <class T>
+// using _my_method_archetype_2 = decltype(declval<T>().my_method());
+KOKKOS_DECLARE_DETECTION_ARCHETYPE_1PARAM(_my_method_archetype_2, T,
+                                          decltype(declval<T>().my_method()));
+
+KOKKOS_STATIC_TEST(is_detected<_my_method_archetype_1, MyTestCase1>::value);
+
+KOKKOS_STATIC_TEST(is_detected<_my_method_archetype_2, MyTestCase1>::value);
+
+KOKKOS_STATIC_TEST(
+    is_detected_exact<void, _my_method_archetype_1, MyTestCase1>::value);
+
+KOKKOS_STATIC_TEST(
+    is_detected_exact<void, _my_method_archetype_2, MyTestCase1>::value);
+
+// KOKKOS_STATIC_TEST_DETECT_EXPRESSION_T(
+//  MyTestCase1,
+//  declval<T>().my_method_with_overloads(declval<C>())
+//);
+//
+// KOKKOS_STATIC_TEST_DETECT_EXPRESSION_T_U(
+//  MyTestCase1, C,
+//  declval<T>().my_method_with_overloads(declval<U>())
+//);
+
+template <class U>
+struct OuterClass {
+  // Things like this don't work with intel or cuda (probably a bug in the EDG
+  // frontend)
```

Comment:
Ok if we are reaching regions where we say it doesn't work because of probably a bug in EDG then we wrote to complicated C++

Reply:
We have lots of other workarounds like this throughout Kokkos.  This is an idiom with widespread public use across C++ for more than 5 years, not some niche corner of C++ trivia.

Reply:
See above, we wanted to use it in a variety of places in RAJA/CAMP for exactly @dhollman's reason, and @wilkill and I came up with a portable version that works around the old nvcc pack expansion and using-directive textual substitution issues.  If you manage to break it, let me know, thusfar we have yet to find a compiler we care about that it fails on.

Reply:
@trws Does this case work with your version? I have more esoteric versions that break more things, but this is the minimal reproducer that I've gotten to fail in quite a few places.  I'd be really happy to be wrong here, though!

Reply:
Oddly enough, the answer isn't entirely black and white.  The default version doesn't pass those as written, a slightly nastier version that I had on the back burner passes two of them on nvcc, icc and pgi, but not the third without modifying the test.  If the checks are defined in the kinda weird concept style we use in camp, you can get something that does the third one, but it would be a re-formulation rather than direct support for nested detect_t on EDG.

Out of curiosity, what cases do you have for this kind of indirection? We usually group our checks into a decltype with a wrapping function, then bind those to a struct or alias to do parametrization of the whole set as a sort of pseudo-concept, but I don't think I've ever run into trying to do this kind of nested dependent detection.  It actually might be nice to find a time to chat in person sometime @dhollman, would be pretty nice if we could come up with a solid option we could use on both sides.  Maybe a good first item for DESUL even.